src/Core/Billing/Commands/Implementations/CancelSubscriptionCommand.cs

```diff
@@ -0,0 +1,149 @@
+﻿using Bit.Core.Billing.Models;
+using Bit.Core.Entities;
+using Bit.Core.Exceptions;
+using Bit.Core.Services;
+using Microsoft.Extensions.Logging;
+using Stripe;
+
+namespace Bit.Core.Billing.Commands.Implementations;
+
+public class CancelSubscriptionCommand : ICancelSubscriptionCommand
+{
+    private readonly ILogger<CancelSubscriptionCommand> _logger;
+    private readonly IStripeAdapter _stripeAdapter;
+
+    private static readonly List<string> _validReasons = new()
+    {
+        "customer_service",
+        "low_quality",
+        "missing_features",
+        "other",
+        "switched_service",
+        "too_complex",
+        "too_expensive",
+        "unused"
+    };
+
+    public CancelSubscriptionCommand(
+        ILogger<CancelSubscriptionCommand> logger,
+        IStripeAdapter stripeAdapter)
+    {
+        _logger = logger;
+        _stripeAdapter = stripeAdapter;
+    }
+
+    public async Task CancelSubscription(
+        ISubscriber subscriber,
+        OffboardingSurveyResponse offboardingSurveyResponse)
+    {
+        var subscription = await GetSubscriptionAsync(subscriber);
+
+        if (IsInactive(subscription))
+        {
+            _logger.LogWarning("Cannot cancel subscription ({ID}) that's already inactive.", subscription.Id);
+            return;
```

Comment:
If we don't throw an `Exception` here then on the `Controller` its considered as a success and then it sends the event to `ReferenceEventService`, correct?

Reply:
@r-tome Good catch; I think this should throw an exception.